Sources/Atomics/AtomicOptionalRawRepresentable.swift

```diff
@@ -0,0 +1,134 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Atomics open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+extension RawRepresentable where RawValue: AtomicOptionalWrappable {
+  public typealias AtomicOptionalRepresentation = RawValue.AtomicOptionalRepresentation
+}
+
+/// A default atomic storage representation for a `RawRepresentable` type
+/// whose `RawValue` conforms to `AtomicOptionalWrappable`.
+@frozen
+public struct AtomicOptionalRawRepresentableStorage<Wrapped>: AtomicStorage
+  where Wrapped: RawRepresentable,
+        Wrapped.RawValue: AtomicOptionalWrappable {
+
+  public typealias Value = Optional<Wrapped>
+
+  @usableFromInline
+  typealias Storage = Wrapped.RawValue.AtomicOptionalRepresentation
+
+  @usableFromInline
+  var _storage: Storage
+
+  @_transparent @_alwaysEmitIntoClient
+  public init(_ value: __owned Optional<Wrapped>) {
+    _storage = Storage(value?.rawValue)
+  }
+
+  @_transparent @_alwaysEmitIntoClient
+  __consuming public func dispose() -> Optional<Wrapped> {
+    _storage.dispose().flatMap(Wrapped.init(rawValue:))
+  }
+
+  @usableFromInline
+  @_transparent @_alwaysEmitIntoClient
+  static func _extract(
+    _ ptr: UnsafeMutablePointer<Self>
+  ) -> UnsafeMutablePointer<Storage> {
+    return UnsafeMutableRawPointer(ptr).assumingMemoryBound(to: Storage.self)
```

Comment:
Nit: we should have the standard comment here explaining why this is okay.

```suggestion
    // `Self` is layout-compatible with its only stored property.
    return UnsafeMutableRawPointer(ptr).assumingMemoryBound(to: Storage.self)
```

Reply:
Good point.